salt/modules/boto_s3.py

```diff
@@ -0,0 +1,176 @@
+
+        s3.region: us-east-1
+
+    If a region is not specified, the default is us-east-1.
+
+    It's also possible to specify key, keyid and region via a profile, either
+    as a passed in dict, or as a string to pull from pillars or minion config:
+
+    .. code-block:: yaml
+
+        myprofile:
+            keyid: GKTADJGHEIQSXMKKRBJ08H
+            key: askdjghsdfjkghWupUjasdflkdfklgjsdfjajkghs
+            region: us-east-1
+
+:depends: boto3
+'''
+# keep lint from choking on _get_conn and _cache_id
+# pylint: disable=E0602
+
+# Import Python libs
+from __future__ import absolute_import
+import logging
+
+# Import Salt libs
+from salt.utils.versions import LooseVersion as _LooseVersion
+
+log = logging.getLogger(__name__)
+
+# pylint: disable=import-error
+try:
+    # pylint: disable=unused-import
+    import boto3
+    # pylint: enable=unused-import
+    import botocore
+    logging.getLogger('boto3').setLevel(logging.CRITICAL)
+    HAS_BOTO = True
+except ImportError:
+    HAS_BOTO = False
+# pylint: enable=import-error
+
+
+def __virtual__():
+    '''
+    Only load if boto libraries exist and if boto libraries are greater than
+    a given version.
+    '''
+    needed_boto3_version = '1.2.1'
+    msg = 'The boto_s3 module cannot be loaded: {0}.'
+    if not HAS_BOTO:
+        return (False, msg.format('boto3 libraries not found'))
+    if _LooseVersion(boto3.__version__) < _LooseVersion(needed_boto3_version):
+        submsg = 'boto3 library version {0} is required'.format(
+            needed_boto3_version,
+        )
+        return (False, msg.format(submsg))
+    return True
+
+
+def __init__(opts):  # pylint: disable=unused-argument
+    if HAS_BOTO:
+        __utils__['boto3.assign_funcs'](__name__, 's3')
+
+
+def get_object_metadata(
+    name,
+    extra_args=None,
+    region=None,
+    key=None,
+    keyid=None,
+    profile=None,
+):
+    '''
+    Get metadata about an S3 object.
+    Returns None if the object does not exist.
+
+    You can pass AWS SSE-C related args and/or RequestPayer in extra_args.
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt myminion boto_s3.get_object_metadata \\
+                         my_bucket/path/to/object \\
+                         region=us-east-1 \\
+                         key=key \\
+                         keyid=keyid \\
+                         profile=profile \\
+    '''
+    bucket, _, key = name.partition('/')
+    if extra_args is None:
+        extra_args = {}
+
+    conn = _get_conn(region=region, key=key, keyid=keyid, profile=profile)
+
+    try:
+        metadata = conn.head_object(
+            Bucket=bucket,
+            Key=key,
+            **extra_args
+        )
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Message'] == 'Not Found':
+            return {'result': None}
+        return {'error': __utils__['boto3.get_error'](e)}
+
+    return {'result': metadata}
+
+
+def upload_object(
```

Comment:
Your implementation of upload_object() seems good, nice and straightforward.  Wondering if there would be any benefit to using `put_object()` rather than `upload_file()` - the former provides a lot more options when putting the object (setting Tags, StorageClass, Metadata, etc, etc.), while the latter is very easy to implement :)  I can definitely see both sides here.  I would recommend that, if you do elect to use `upload_file()` for this function, that the function itself be called `upload_file()` rather than `upload_object()` since the latter might confuse someone, and also makes referencing the relevant AWS API docs or boto3 pages a bit harder.  In general I think it's good practice to name execution functions, where possible, to match whichever boto3 functions they are calling under the hood.

Reply:
As far as I understand, both APIs provide the same options, but `upload_file` just bundles them all under the `ExtraArgs` kwarg, so I think that makes the code a bit simpler here.

I'll rename this to `upload_file`, makes sense.